//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "ni.niScope";
option java_outer_classname = "niScope";
option objc_class_prefix = "NI";

option cc_enable_arenas = true;

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package niScope;

//---------------------------------------------------------------------
// The niScope service definition.
//---------------------------------------------------------------------
service niScopeService {
    rpc StreamLatencyTest(stream StreamLatencyClient) returns (stream StreamLatencyServer) {};
    rpc StreamLatencyTestClient(stream StreamLatencyClient) returns (StreamLatencyServer) {};
    rpc StreamLatencyTestServer(StreamLatencyClient) returns (stream StreamLatencyServer) {};
    rpc Init(InitParameters) returns (InitResult) {};
    rpc InitWithOptions(InitWithOptionsParameters) returns (InitWithOptionsResult) {};
    rpc AutoSetup(AutoSetupParameters) returns (AutoSetupResult) {};
    rpc ConfigureHorizontalTiming(ConfigureHorizontalTimingParameters) returns (ConfigureHorizontalTimingResult) {};
    rpc TestWrite(TestWriteParameters) returns (TestWriteResult) {};
    rpc TestWriteContinuously(stream TestWriteParameters) returns (stream TestWriteResult) {};
    rpc Read(ReadParameters) returns (ReadResult) {};
    rpc ReadContinuously(ReadContinuouslyParameters) returns (stream ReadContinuouslyResult) {};
}

message StreamLatencyClient {
  uint32 message = 1;
}

message StreamLatencyServer {
  uint32 message = 1;
}

message ViSession {
  uint32 id = 1;
}

message InitParameters {
  int32 id = 1;
}
    
message InitResult {
  int32 status = 1;
}

message InitWithOptionsParameters {
  string resourceName = 1;
  bool IDQuery = 2;
  bool resetDevice = 3;
  string optionString = 4;
}
    
message InitWithOptionsResult {
  int32 status = 1;
  ViSession newVi = 2;
}

message AutoSetupParameters {
  ViSession vi = 1;
}
    
message AutoSetupResult {
  int32 status = 1;
}

message ConfigureHorizontalTimingParameters {
  ViSession vi = 1;
  double minSampleRate = 2;
  int32 minNumPts = 3;
  double refPosition = 4;
  int32 numRecords = 5;
  bool enforceRealtime = 6;
}
    
message ConfigureHorizontalTimingResult {
  int32 status = 1;
}
message ReadParameters {
  ViSession vi = 1;
  string channelList = 2;
  double timeout = 3;
  int32 numSamples = 4;
}
    
message ReadResult {
  int32 status = 1;
  repeated double wfm = 5;
  repeated ScopeWaveformInfo wfmInfo = 2;
}

message TestWriteParameters {
  ViSession vi = 1;
  repeated double wfm = 2;
}
    
message TestWriteResult {
  int32 status = 1;
}

message ReadContinuouslyParameters {
  int32 numIterations = 1;
  int32 numSamples = 2;
}
    
message ReadContinuouslyResult {
  int32 status = 1;
  repeated double wfm = 2;
}

message ScopeWaveformInfo {
  double absoluteInitialX = 1;
  double relativeInitialX = 2;
  double xIncrement = 3;
  int32 actualNumberOfSamples = 4;
  double offset = 5;
  double gain = 6;
}
